libraries/standard/mqtt/utest/mqtt_utest.c

```diff
@@ -78,3 +142,546 @@ void test_MQTT_Init_Invalid_Params( void )
+    if( expectMoreCalls )
+    {
+        if( incomingPublish )
+        {
+            MQTT_UpdateStatePublish_ExpectAnyArgsAndReturn( stateAfterDeserialize );
+        }
+        else
+        {
+            MQTT_UpdateStateAck_ExpectAnyArgsAndReturn( stateAfterDeserialize );
+        }
+
+        if( stateAfterDeserialize == MQTTPublishDone )
+        {
+            expectMoreCalls = false;
+        }
+    }
+
+    /* Serialize the packet to be sent in response to the received packet.
+     * Observe that there is no reason to serialize a PUBLISH after receiving
+     * a packet. */
+    if( expectMoreCalls )
+    {
+        MQTT_SerializeAck_ExpectAnyArgsAndReturn( serializeStatus );
+
+        if( serializeStatus != MQTTSuccess )
+        {
+            expectMoreCalls = false;
+        }
+    }
+
+    /* Update the state based on the sent packet. */
+    if( expectMoreCalls )
+    {
+        MQTT_UpdateStateAck_ExpectAnyArgsAndReturn( stateAfterSerialize );
+    }
+
+    /* Expect the above calls when running MQTT_ProcessLoop. */
+    mqttStatus = MQTT_ProcessLoop( pContext, MQTT_NO_TIMEOUT_MS );
+    TEST_ASSERT_EQUAL( processLoopStatus, mqttStatus );
+
+    /* Any final assertions to end the test. */
+    if( mqttStatus == MQTTSuccess )
+    {
+        if( currentPacketType == MQTT_PACKET_TYPE_PUBLISH )
+        {
+            TEST_ASSERT_TRUE( pContext->controlPacketSent );
+        }
+
+        if( currentPacketType == MQTT_PACKET_TYPE_PINGRESP )
+        {
+            TEST_ASSERT_FALSE( pContext->waitingForPingResp );
+        }
+    }
+}
+
+/**
+ * @brief Test that NULL pContext causes MQTT_ProcessLoop to return MQTTBadParameter.
+ */
+void test_MQTT_ProcessLoop_Invalid_Params( void )
+{
+    MQTTStatus_t mqttStatus = MQTT_ProcessLoop( NULL, MQTT_NO_TIMEOUT_MS );
+
+    TEST_ASSERT_EQUAL( MQTTBadParameter, mqttStatus );
+}
+
+/**
+ * @brief This test case covers all calls to the private method,
+ * handleIncomingPublish(...),
+ * that results in the process loop returning successfully.
+ */
+void test_MQTT_ProcessLoop_handleIncomingPublish_Happy_Paths( void )
+{
+    MQTTStatus_t mqttStatus;
+    MQTTContext_t context;
+    MQTTTransportInterface_t transport;
+    MQTTFixedBuffer_t networkBuffer;
+    MQTTApplicationCallbacks_t callbacks;
+
+    setupTransportInterface( &transport );
+    setupCallbacks( &callbacks );
+    setupNetworkBuffer( &networkBuffer );
+
+    mqttStatus = MQTT_Init( &context, &transport, &callbacks, &networkBuffer );
+    TEST_ASSERT_EQUAL( MQTTSuccess, mqttStatus );
+
+    modifyIncomingPacketStatus = MQTTSuccess;
+
+    /* Assume QoS = 1 so that a PUBACK will be sent after receiving PUBLISH. */
+    currentPacketType = MQTT_PACKET_TYPE_PUBLISH;
+    expectProcessLoopCalls( &context, MQTTSuccess, MQTTPubAckSend,
+                            MQTTSuccess, MQTTPublishDone,
+                            MQTTSuccess, true );
+
+    /* Assume QoS = 2 so that a PUBREC will be sent after receiving PUBLISH. */
```

Comment:
Ditto?

Reply:
`expectProcessLoopCalls` has `incomingPublish=1` and `stateAfterDeserialize=MQTTPubRecSend`. It's just a matter of changing `stateAfterDeserialize` depending on assumed QoS.